File: swift-bridge/Sources/SpriteKitBridge/AudioNode.swift
import AVFoundation
import SpriteKit

@_cdecl("sk_audio_node_new")
public func sk_audio_node_new() -> UnsafeMutableRawPointer? {
    skRetain(SKAudioNode(avAudioNode: AVAudioPlayerNode()))
}

@_cdecl("sk_audio_node_get_autoplay_looped")
public func sk_audio_node_get_autoplay_looped(_ nodeHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SKAudioNode = skBorrow(nodeHandle) else { return true }
    return node.autoplayLooped
}

@_cdecl("sk_audio_node_set_autoplay_looped")
public func sk_audio_node_set_autoplay_looped(_ nodeHandle: UnsafeMutableRawPointer?, _ autoplay: Bool) {
    guard let node: SKAudioNode = skBorrow(nodeHandle) else { return }
    node.autoplayLooped = autoplay
}

@_cdecl("sk_audio_node_get_positional")
public func sk_audio_node_get_positional(_ nodeHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SKAudioNode = skBorrow(nodeHandle) else { return false }
    return node.isPositional
}

@_cdecl("sk_audio_node_set_positional")
public func sk_audio_node_set_positional(_ nodeHandle: UnsafeMutableRawPointer?, _ positional: Bool) {
    guard let node: SKAudioNode = skBorrow(nodeHandle) else { return }
    node.isPositional = positional
}

@_cdecl("sk_audio_node_has_av_audio_node")
public func sk_audio_node_has_av_audio_node(_ nodeHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SKAudioNode = skBorrow(nodeHandle) else { return false }
    return node.avAudioNode != nil
}

File: src/audio_node.rs
use crate::ffi;
use crate::node::AsNode;
use crate::private::handle_type;

handle_type!(AudioNode);

impl AsNode for AudioNode {
    fn as_node_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

impl AudioNode {
    /// Wraps `SKAudioNode`.
    #[must_use]
    pub fn new() -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_audio_node_new()) }
    }

    /// Returns a property exposed by `SKAudioNode`.
    #[must_use]
    pub fn autoplay_looped(&self) -> bool {
        unsafe { ffi::sk_audio_node_get_autoplay_looped(self.ptr) }
    }

    /// Sets a property exposed by `SKAudioNode`.
    pub fn set_autoplay_looped(&self, autoplay: bool) {
        unsafe { ffi::sk_audio_node_set_autoplay_looped(self.ptr, autoplay) };
    }

    /// Returns a property exposed by `SKAudioNode`.
    #[must_use]
    pub fn is_positional(&self) -> bool {
        unsafe { ffi::sk_audio_node_get_positional(self.ptr) }
    }

    /// Sets a property exposed by `SKAudioNode`.
    pub fn set_positional(&self, positional: bool) {
        unsafe { ffi::sk_audio_node_set_positional(self.ptr, positional) };
    }

    /// Returns a property exposed by `SKAudioNode`.
    #[must_use]
    pub fn has_audio_node(&self) -> bool {
        unsafe { ffi::sk_audio_node_has_av_audio_node(self.ptr) }
    }
}

impl Default for AudioNode {
    fn default() -> Self {
        Self::new().expect("SKAudioNode creation failed")
    }
}
